Expose Typescript types in libSquoosh

ergunsh opened the pull request:
This PR and https://github.com/GoogleChromeLabs/squoosh/pull/1141 should fix https://github.com/GoogleChromeLabs/squoosh/issues/1077

Essentially, we're creating the declaration files as well in the TS build process and moving them to the libSquoosh build directory. Then, we expose the types through package.json's `types` field.

ergunsh committed:
Expose type declarations in libSquoosh npm package

ergunsh committed:
Add comment on why we remove the tsbuildinfo

surma commented:
Thanks @ergunsh!

I actually think that @jakearchibald’s `simple-ts` rollup plugin is capable of emitting `.d.ts`. Did you try that? @jakearchibald do you know what would be necessary?

ergunsh commented:
Hey hey, in here actually simple-ts plugin emits the `.d.ts` files. However, our output folder for tsc is `../.tmp` but our build files and the files we publish through npm are in `libSquoosh/build`; that's why we needed the extra copy step at the end.

If there is a way to do it automatically with simple-ts, it would be great! :)

We can try to update the output folder for tsc from `../.tmp` to `./build` but it also means we'll have some unnecessary `.js` files in the build folder if we don't remove them explicitly before publishing.

surma reviewed (approved):
Let’s move forward with this as is :)

surma commented:
@ergunsh I merged your other PR. Do you mind updating the branch and checking that everything still works? Then we can merge this.

ergunsh committed:
Merge branch 'dev' into expose-types

ergunsh committed:
Fix PreprocessOptions type

Resize should require at least one of the width, height.
The other options are optional for all preprocessors

ergunsh committed:
Update libSquoosh README to reflect encode changes

I also removed requiring `await image.decoded` call before calling
preprocess or encode since they decode the image before the operation

ergunsh commented:
Hey @surma, sorry for the late action 😅

I fixed one thing before exposing the types: PreprocessOptions were requiring all the options but preprocessors don't require that.

In addition to that, I updated README to reflect encode changes.

After these changes everything seems to be working fine from typing perspective :)

ergunsh commented:
Hey @surma, friendly ping 🏓 :)

ergunsh committed:
Merge branch 'dev' into expose-types

ergunsh committed:
Merge branch 'dev' into expose-types

aprilmintacpineda commented:
We just had this error, we had to add `// @ts-ignore` above the `import` line. Can someone please help merge this? Thanks very much! 🙏

Dandarprox commented:
**PING** Please merge this :pray:

jakearchibald commented:
LGTM! Sorry for the delay on this.

jakearchibald commented:
Published as 0.4.1

donmccurdy commented:
Just FYI for others tracking this – it looks like 0.4.1 was published and then reverted, see https://github.com/GoogleChromeLabs/squoosh/issues/1223 for details.
